application/ledger_query_service/src/ethereum/queries.rs

```diff
@@ -142,45 +173,85 @@ fn clean_0x(s: &str) -> &str {
     }
 }
 
-impl LogQuery {
+impl EventQuery {
     pub fn matches_block(&self, block: &Block<Transaction>) -> bool {
         match self {
-            Self { topics, .. } if topics.is_empty() => false,
-            Self { topics } => topics.iter().all(|topics| {
-                topics
-                    .iter()
-                    .all(|topic| block.logs_bloom.contains_input(Input::Raw(&topic)))
+            Self { event_matchers } if event_matchers.is_empty() => false,
+            Self { event_matchers } => event_matchers.iter().all(|log_matcher| {
+                log_matcher.topics.iter().all(|topic| {
+                    if let Some(topic) = topic {
+                        block.logs_bloom.contains_input(Input::Raw(&topic))
+                    } else {
+                        false
+                    }
+                })
             }),
         }
     }
 
     pub fn matches_transaction_receipt(&self, transaction_receipt: TransactionReceipt) -> bool {
         match self {
-            Self { topics } if topics.is_empty() => false,
-            Self { topics } => topics.iter().all(|topics| {
-                !topics.is_empty()
-                    && transaction_receipt
-                        .logs
-                        .iter()
-                        .any(|tx_log| topics.iter().all(|topic| tx_log.topics.contains(topic)))
-            }),
+            Self { event_matchers } if event_matchers.is_empty() => false,
+            Self { event_matchers } => {
+                event_matchers
+                    .iter()
+                    .all(|event_matcher| match event_matcher {
+                        EventMatcher {
+                            address: None,
+                            data: None,
+                            topics,
+                        } if topics.is_empty() => false,
+                        EventMatcher {
+                            address,
+                            data,
+                            topics,
+                        } => {
+                            transaction_receipt.logs.iter().any(|tx_log| {
+                                let mut result = true;
```

Comment:
This looks messy but there is probably not much benefit in optimizing it further :)

We could make a `good-first-issue` for refactoring tasks like that!

Reply:
@comit-network/rust-devs what do you think about that?

Reply:
leave it like that for now™

Reply:
I tried something fancy like this but ended up in compiler errors _moving_ blabla_ into _FnMut_ blabla so I let it go :D   
```
if data.map_or(false, |data| data == txlog.data) {
return false;
}
```

Reply:
ok, thanks to @thomaseizinger this worked now.